### PR TITLE
python3Packages.certifi: Update and use system ca-bundle

### PR DESCRIPTION
**Pull request by mweinelt**
We update the system ca-bundle more reliably, and it allows ties in with module based
configuration applied through `security.pki`.

Fixes: CVE-2022-23491
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vcunat** [comment]
@OfBorg eval

File: pkgs/development/python-modules/certifi/default.nix [hunk 1]
```diff
@@ -1,23 +1,30 @@
...
     repo = "python-certifi";
     rev = version;
-    hash = "sha256-B6LO6AfG9cfpyNI7hj3VjmGTFsrrIkDYO4gPMkZY74w=";
+    hash = "sha256-r6TJ6YGL0cygz+F6g6wiqBfBa/QKhynZ92C6lHTZ2rI=";
   };
 
+  postPatch = ''
+    # Use our systemwide ca-bundle
+    substituteInPlace certifi/core.py \
+      --replace "return _CACERT_PATH" 'return "${cacert}/etc/ssl/certs/ca-bundle.crt"'
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/certifi/default.nix:25]: As this is quite critical to still apply in the future I would think we would want to use a patch to know that it is always applied.

Comment (mweinelt) [reply pkgs/development/python-modules/certifi/default.nix:25]: Good idea.

Comment (SuperSandro2000) [reply pkgs/development/python-modules/certifi/default.nix:25]: Also codewise maybe replacing https://github.com/certifi/python-certifi/blob/2022.12.07/certifi/core.py#L14-L15 would be cleaner

**Review by SuperSandro2000 — APPROVED** [review]
didn't double check the patch with the original code but LGTM

**Comment by vcunat** [comment]
This broke the build of `python3.pkgs.sphinx`:
https://hydra.nixos.org/log/mlnd1mlnk9gr9zg79p5yqxbdkyf82mvq-python3.10-sphinx-5.1.1.drv
(I checked that reverting this merge fixes it.)

**Comment by mweinelt** [comment]
Happens because of the dummy NIX_SSL_CERT_FILE path passed in from `stdenv/generic/steup.sh`. Two solutions two this.

- Unset `NIX_SSL_CERT_FILE` in `preCheck`
- Pass `cacert` into `checkInputs`, because its setup script sets up the variable correctly

I think unsetting is fine.

**Comment by vcunat** [comment]
I wonder if `certifi` should fix this automatically, e.g. simply propagate `cacert` when it's in runtime closure anyway.

**Comment by vcunat** [comment]
Let me elaborate.  My understanding of the current design:
- TLS-related libs prefer not to have any CA certs in closure and during runtime default to OS's current bundle (impure... similarly to the real date progressing impurely)
- in case certs are needed in tests (not common because they don't have internet access), they put `cacert` into inputs
- I'd expect that `certifi` will be used similarly to `cacert`

**Comment by vcunat** [comment]
Hmm, apparently it didn't cover all the cases.  I hoped it would work better.

Right now I see `pytest-param-files`, `more-itertools`, `google-auth` and there will probably be more.  Any idea about a more general fix, or will we add cacert or certifi as some input to each of them?

**Comment by trofi** [comment]
Bisect claims that 8d7cc9cac9ecdf95f554c5ea7ca15118baa06c39 `python3Packages.certifi: use system ca-bundle` broke `python3Packages.scikit-build` tests in `staging` and `staging-next`:

```
$ nix build --no-link -f. -L python3Packages.scikit-build
...
python3.10-scikit-build> ==================================== ERRORS ====================================
python3.10-scikit-build> _____________________ ERROR collecting tests/test_setup.py _____________________
python3.10-scikit-build> tests/test_setup.py:329: in <module>
python3.10-scikit-build>     @pytest.mark.skipif(not is_site_reachable("https://pypi.org/simple/cmake/"), reason="pypi.org website not reachable")
python3.10-scikit-build>         CMAKE_INSTALL_DIR = <function CMAKE_INSTALL_DIR at 0x7ffff4b30280>
python3.10-scikit-build>         SKBUILD_DIR = <function SKBUILD_DIR at 0x7ffff4b300d0>
python3.10-scikit-build>         SKBuildError = <class 'skbuild.exceptions.SKBuildError'>
python3.10-scikit-build>         __builtins__ = <builtins>
python3.10-scikit-build>         __cached__ = '/build/scikit-build-0.15.0/tests/__pycache__/test_setup.cpython-310.pyc'
python3.10-scikit-build>         __doc__    = 'test_setup\n----------------------------------\n\nTests for `skbuild.setup` function.\n'
python3.10-scikit-build>         __file__   = '/build/scikit-build-0.15.0/tests/test_setup.py'
python3.10-scikit-build>         __loader__ = <_pytest.assertion.rewrite.AssertionRewritingHook object at 0x7ffff6a40a60>
python3.10-scikit-build>         __name__   = 'tests.test_setup'
python3.10-scikit-build>         __package__ = 'tests'
...
```

**Comment by mweinelt** [comment]
Annoying. The most simple fix would be to ignore `NIX_SSL_CERT_FILE` when set to `/no-cert-file.crt`, so we'll force to use the symlinked ca-bundle from cacert.

**Pull request merged**